File: c/lib/PIcontroller/PIcontroller.c
#include "PIcontroller.h"

void pi_controller_init(PIController *pid,
                                       double Kp, double Ki, double Ts,
                                       bool has_min, double output_min,
                                       bool has_max, double output_max)
{
    pid->Kp = Kp;
    pid->Ki = Ki;
    pid->Ts = Ts;

    pid->has_min = has_min;
    pid->has_max = has_max;
    pid->output_min = has_min ? output_min : PI_NO_LIMIT_MIN;
    pid->output_max = has_max ? output_max : PI_NO_LIMIT_MAX;

    pid->integral  = 0.0;
}

void pi_controller_reset(PIController *pid)
{
    pid->integral = 0.0;
}

double pi_controller_update(PIController *pid,
                                           double reference, double feedback)
{
    pid->error = reference - feedback;
    double proportional = pid->Kp * pid->error;

    pid->integral += pid->Ki * pid->error * pid->Ts;
    double output = proportional + pid->integral;


    if (pid->has_min && output < pid->output_min) {
        output = pid->output_min;
        if (pid->error < 0.0) {
            pid->integral -= pid->Ki * pid->error * pid->Ts;
        }
    }

    if (pid->has_max && output > pid->output_max) {
        output = pid->output_max;
        if (pid->error > 0.0) {
            pid->integral -= pid->Ki * pid->error * pid->Ts;
        }
    }

    return output;
}

File: python/bldc_python/PIcontroller.py
import numpy as np
from numpy import pi

class PIController:

    def __init__(self, Kp, Ki, Ts, output_min=None, output_max=None):
        self.Kp = Kp
        self.Ki = Ki
        self.Ts = Ts
        self.output_min = output_min
        self.output_max = output_max
        self.integral = 0.0
        self.saturated = False  # instrumentacao para diagnostico

    def reset(self):
        self.integral = 0.0

    def update(self, reference, feedback):
        error = reference - feedback
        proportional = self.Kp * error
        self.integral += self.Ki * error * self.Ts
        output = proportional + self.integral

        self.saturated = False

        if self.output_min is not None:
            if output < self.output_min:
                output = self.output_min
                self.saturated = True
                if error < 0:
                    self.integral -= self.Ki * error * self.Ts

        if self.output_max is not None:
            if output > self.output_max:
                output = self.output_max
                self.saturated = True
                if error > 0:
                    self.integral -= self.Ki * error * self.Ts

        return output